### PR TITLE
feat: Repartition - support ArrayList

### PR DESCRIPTION
**Pull request by gandronchik**
Repartition Physical Node does not support ArrayList. This PR solves this problem

### CODE REVIEW

File: datafusion/core/src/physical_plan/hash_utils.rs [hunk 1]
```diff
@@ -257,6 +258,32 @@ fn create_hashes_dictionary<K: ArrowDictionaryKeyType>(
...
+    array: &ArrayRef,
+    random_state: &RandomState,
+    hashes_buffer: &mut [u64],
+) -> Result<()> {
+    let list_arr = array
+        .as_any()
+        .downcast_ref::<GenericListArray<T>>()
+        .unwrap();
+
+    let mut values: Vec<ArrayRef> = Vec::new();
```
Comment (xudong963) [inline datafusion/core/src/physical_plan/hash_utils.rs:272]:
```suggestion
    let mut values: Vec<ArrayRef> = Vec::with_capacity(list_arr.len());
```

**Review by alamb — APPROVED** [review]
This code looks reasonable to me. Thank you @gandronchik 

This PR looks like it has a clippy failure and a comment from @xudong963 but otherwise I think it is good to go.

**Comment by alamb** [comment]
I apologize for the delay in review -- please let us know how you would like to proceed with this PR

**Comment by alamb** [comment]
Marking as draft as this PR is waiting on a response from @gandronchik

**Comment by alamb** [comment]
Please mark it ready for review when it is ready

**Comment by alamb** [comment]
Closing as this PR is over a year old. Please feel free to reopen it / rebase it if you plan to keep working on it

**Pull request closed without merging**